SparseGrids/tsgRuleWavelet.cpp

```diff
@@ -390,51 +415,77 @@ double RuleWavelet::getSupport(int point) const{
     if (subindex == 0){
-        return linear_boundary_wavelet(scale * (x + 1.) - 1.);
-    }
-    // Right Boundary
-    else if (subindex == (1 << l) - 1){
-        return linear_boundary_wavelet(scale * (1. - x) - 1.);
+        // Left boundary wavelet.
+        value = linear_boundary_wavelet<mode>(scale * (x + 1.) - 1., x <= 0.0);
+    } else if (subindex == (1 << l) - 1) {
+        // Right boundary wavelet.
+        value = linear_boundary_wavelet<mode>(scale * (1. - x) - 1., x > 0.0);
+    } else {
+        // Central wavelets
+        double shift = 0.5 * (double (subindex - 1));
+        value = linear_central_wavelet<mode>(scale * (x + 1) - 1. - shift, x <= 0.0);
     }
-    double shift = 0.5 * (double (subindex - 1));
-    return linear_central_wavelet(scale * (x + 1) - 1. - shift);
+    // Adjust for the chain rule multiplier.
+    if (mode == 1) value *= (subindex == (1 << l) - 1) ? -scale : scale;
+
+    return value;
 }
 
-inline double RuleWavelet::linear_boundary_wavelet(double x) const{
-    // Evaluates the first order boundary wavelet with support on [-1, 0].
+template<int mode>
+inline double RuleWavelet::linear_boundary_wavelet(double x, bool right) const{
+    // Evaluates or differentiates the first order boundary wavelet with support on [-1, 0].
+    // If `right` is true, then we return the right derivatives/values (if they exist); else, we return the left ones.
     if (std::abs(x + 0.5) > 0.5) return 0.0;
 
-    if ((x <= -0.75)){
-        return 0.75 * (7.0 * x + 6.0);
-    }else if (x <= -0.5){
-        return -0.25 * (11.0 * x + 6.0);
-    }else{
-        return 0.25 * x;
+    if ((right and x < -0.75) or (!right and x <= -0.75)) {
+        if (mode == 0) return 0.75 * (7.0 * x + 6.0);
+        else return 0.75 * 7.0;
+    } else if ((right and x < -0.5) or (!right and x <= -0.5)) {
+        if (mode == 0) return -0.25 * (11.0 * x + 6.0);
+        else return -0.25 * 11.0;
+    } else if ((right and x < 0) or (!right and x <= 0)) {
+        if (mode == 0) return 0.25 * x;
+        else return 0.25;
+    } else {
+        // Case: (right and x == 0.0)
+        return 0.0;
     }
 }
 
-inline double RuleWavelet::linear_central_wavelet(double x) const {
-    // Evaluates the first order central wavelet with support on [-1, .5].
+template<int mode>
+inline double RuleWavelet::linear_central_wavelet(double x, bool right) const {
+    // Evaluates or differentiates (right derivative only) the first order central wavelet with support on [-1, .5].
+    // If `right` is true, then we return the right derivatives/values (if they exist); else, we return the left ones.
     if (std::abs(x + 0.25) > 0.75) return 0.0;
 
-    if ((x <= -0.5)){
-        return -0.5 * x - 0.5;
-    }else if (x >= 0.0){
-        return 0.5 * x - 0.25;
-    }else if (x <= -0.25){
-        return 4.0 * x + 1.75;
-    }else{
-        return -4.0 * x - 0.25;
+    // The ordering is important.
+    if ((right and x < -0.5) or (!right and x <= -0.5)) {
```

Comment:
This is a bit messy, the `right` and `!right` are almost identical. How about:
```
if (x < -0.5 or (x == -0.5 and not right))
```
I also don't like that the messy right-not-right logic will come into play in both modes. I would like to have one section for `mode == 0` that does the old logic without the messy `!right` and a separate differentiation section that does the special case work.